src/validation.cpp

```diff
@@ -1061,28 +1178,131 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+    return MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+}
+
+PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+
+    PackageValidationState package_state;
+    const unsigned int package_count = txns.size();
+
+    // These context-free package limits can be checked before taking the mempool lock.
+    if (package_count > MAX_PACKAGE_COUNT) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-many-transactions");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    const int64_t total_size = std::accumulate(txns.cbegin(), txns.cend(), 0,
+                               [](int64_t sum, const auto& tx) { return sum + GetVirtualTransactionSize(*tx); });
+    if (total_size > MAX_PACKAGE_SIZE * 1000) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-large");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(package_count);
+
+    // Require the package to be sorted in order of dependency, i.e. parents appear before children.
+    // An unsorted package will fail anyway on missing-inputs, but it's better to quit earlier and
+    // fail on something less ambiguous (missing-inputs could also be an orphan or trying to
+    // spend nonexistent coins).
+    for (const auto& tx : txns) {
+        Workspace workspace(tx);
+        for (const auto& input : tx->vin) {
+            if (std::find_if(workspaces.begin(), workspaces.end(),
+                             [&, txid = input.prevout.hash](const auto& preceding_ws)
+                             { return preceding_ws.m_ptx->GetHash() == txid; }) == workspaces.end() &&
+                std::find_if(txns.begin(), txns.end(),
+                             [&, txid = input.prevout.hash](const auto& tx)
+                             { return tx->GetHash() == txid; }) != txns.end()) {
+                // If we don't find the parent in workspaces but we do find it in txns,
+                // then it must be a subsequent transaction in the package.
+                package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-not-sorted");
+                return PackageMempoolAcceptResult(package_state, {});
+            }
+        }
+        workspaces.emplace_back(std::move(workspace));
+    }
+
+    LOCK(m_pool.cs);
+
+    std::map<const uint256, const MempoolAcceptResult> results;
+    // Do all PreChecks first and fail fast to avoid running expensive script checks when unnecessary.
+    for (Workspace& ws : workspaces) {
+        if (!PreChecks(args, ws)) {
```

Comment:
separately -- is it true that PreChecks is not entirely stateless? On edge condition, 2 separate ATMPs could have different minfees & therefore this package code is a little different than separate submissions.

Reply:
`PreChecks` is not stateless, no. Maybe you're thinking of `CheckTransaction`?
There's no way to make separate mempool accepts identical unless we hold cs_main the entire time - you can just as well have different results from the inputs getting spent in between calls.